VPD parser crashes in some cases (case 2)

# Required information

Distribution: Ubuntu
Distribution version: 22.04 (not specific to distro)
Kernel version: 5.15.0-40-generic (not specific to this kernel)
LXC version: 5.0.0
LXD version: https://github.com/lxc/lxd/commit/8539b2ccc6286cf4ae58eac6b9e455a0f239b0d2


# Issue description

This is another instance of a VPD parser crash as in https://github.com/lxc/lxd/issues/10627 but on a different PCI device (crashes on other devices are no longer triggered).

I attached `/sys/bus/pci/devices/0000:05:00.0/vpd` which was the last vpd file opened in the strace below:
[bad-vpd-3.zip](https://github.com/lxc/lxd/files/9093747/bad-vpd-3.zip)

```
sudo strace -f usr/share/maas/machine-resources/amd64 2>&1 | grep /vpd
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:00.0/vpd",  <unfinished ...>
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:01.0/vpd", 0xc0009e5bd8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:02.0/vpd",  <unfinished ...>
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:03.0/vpd", 0xc000a0c038, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:05.0/vpd", 0xc000a0c788, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:05.1/vpd", 0xc000a0ced8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:05.2/vpd", 0xc000a0d628, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:05.4/vpd", 0xc000a0dd78, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:11.0/vpd", 0xc000a22518, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:11.4/vpd", 0xc000a237c8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:16.0/vpd", 0xc000a23f18, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:16.1/vpd", 0xc000aa46b8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1a.0/vpd", 0xc000aa5898, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1c.0/vpd", 0xc0009b9d78, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1c.4/vpd", 0xc0009c8fa8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1c.7/vpd", 0xc0009d01d8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1d.0/vpd", 0xc0009d13b8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1f.0/vpd", 0xc0009d85e8, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:00:1f.2/vpd", 0xc0009d9898, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:00.0/vpd", {st_mode=S_IFREG|0600, st_size=32768, ...}, AT_SYMLINK_NOFOLLOW) = 0
[pid 187832] openat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:00.0/vpd", O_RDONLY|O_CLOEXEC) = 3
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:00.1/vpd", {st_mode=S_IFREG|0600, st_size=32768, ...}, AT_SYMLINK_NOFOLLOW) = 0
[pid 187832] openat(AT_FDCWD, "/sys/bus/pci/devices/0000:01:00.1/vpd", O_RDONLY|O_CLOEXEC <unfinished ...>
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:02:00.0/vpd",  <unfinished ...>
[pid 187832] openat(AT_FDCWD, "/sys/bus/pci/devices/0000:02:00.0/vpd", O_RDONLY|O_CLOEXEC) = 3
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:03:00.0/vpd", {st_mode=S_IFREG|0600, st_size=32768, ...}, AT_SYMLINK_NOFOLLOW) = 0
[pid 187832] openat(AT_FDCWD, "/sys/bus/pci/devices/0000:03:00.0/vpd", O_RDONLY|O_CLOEXEC) = 3
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:03:00.1/vpd",  <unfinished ...>
[pid 187832] openat(AT_FDCWD, "/sys/bus/pci/devices/0000:03:00.1/vpd", O_RDONLY|O_CLOEXEC) = 3
[pid 187832] newfstatat(AT_FDCWD, "/sys/bus/pci/devices/0000:05:00.0/vpd",  <unfinished ...>
[pid 187832] openat(AT_FDCWD, "/sys/bus/pci/devices/0000:05:00.0/vpd", O_RDONLY|O_CLOEXEC) = 3
```

```
$ sudo usr/share/maas/machine-resources/amd64
panic: runtime error: slice bounds out of range [2:1]

goroutine 1 [running]:
github.com/lxc/lxd/lxd/resources.vpdReadInt(...)
        /build/maas-vsXIsL/maas-3.2.0-11992-g.0fe4c5431/src/host-info/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:44
github.com/lxc/lxd/lxd/resources.parsePCIVPD({0xc000ee0000, 0x8000, 0x8001})
        /build/maas-vsXIsL/maas-3.2.0-11992-g.0fe4c5431/src/host-info/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:94 +0x390
github.com/lxc/lxd/lxd/resources.GetPCI()
        /build/maas-vsXIsL/maas-3.2.0-11992-g.0fe4c5431/src/host-info/vendor/github.com/lxc/lxd/lxd/resources/pci.go:146 +0x9dc
github.com/lxc/lxd/lxd/resources.GetResources()
        /build/maas-vsXIsL/maas-3.2.0-11992-g.0fe4c5431/src/host-info/vendor/github.com/lxc/lxd/lxd/resources/resources.go:48 +0x267
host-info/pkg/info.GetResources(...)
        /build/maas-vsXIsL/maas-3.2.0-11992-g.0fe4c5431/src/host-info/pkg/info/info.go:157
host-info/pkg/info.GetInfo()
        /build/maas-vsXIsL/maas-3.2.0-11992-g.0fe4c5431/src/host-info/pkg/info/info.go:182 +0x34
main.main()
        /build/maas-vsXIsL/maas-3.2.0-11992-g.0fe4c5431/src/host-info/cmd/machine-resources/main.go:23 +0x1d
```

--- a/lxd/resources/pci_vpd.go
+++ b/lxd/resources/pci_vpd.go
@@ -91,6 +91,11 @@ func parsePCIVPD(buf []byte) api.ResourcesPCIVPD {
 		buf, tag = vpdReadInt(buf, 1)
 		if (tag & 0x80) == 0x80 {
 			// Large resource data, Read the 2-bytes entry length.
+
+			if len(buf) < 2 {
+				break
+			}
+
 			buf, length = vpdReadInt(buf, 2)
 		} else {
 			// Small resource data, size is in the tag itself.

